libres/tests/analysis/test_update.cpp

```diff
@@ -0,0 +1,341 @@
+    int ens_size = matrix_get_columns(A);
+    int nparam = matrix_get_rows(A);
+    matrix_type *A_hat = matrix_alloc_copy(A);
+    matrix_subtract_row_mean(A_hat); // Equation (9.13)
+    matrix_type *C_psi_e = matrix_alloc(nparam, nparam);
+    matrix_matmul_with_transpose(C_psi_e, A_hat, A_hat, false, true);
+
+    matrix_free(A_hat);
+
+    matrix_scale(C_psi_e, 1.0 / (ens_size - 1));
+
+    return C_psi_e;
+}
+
+/**
+ * @brief Determinant of a 2 x 2 matrix
+ * 
+ * @param C 
+ * @return double 
+ */
+static double matrix_det(const matrix_type *C) {
+    double det_C = matrix_iget(C, 0, 0) * matrix_iget(C, 1, 1) -
+                   matrix_iget(C, 1, 0) * matrix_iget(C, 0, 1);
+    return det_C;
+}
+
+/*
+# Implementing the Kalman filter as defined in equations (3.83), (3.84) and (3.85)
+# in Data Assimilation: The Ensemble Kalman Filter, Geir Evensen, 2009.
+# This is useful for sanity-checking tests.
+
+# Hints:
+# Setting obs_std very low means high confidence in observations,
+# which means it should only take 1 iteration (niter=1) to get a good estimate.
+# Similarly, a very high obs_std will require lots of iterations.
+
+import numpy as np
+rng = np.random.default_rng(12345)
+
+# The model is a two-parameter linear model.
+nparam = 2
+a_t = 1
+b_t = 5
+
+# Defining priors
+a_f_std = 2
+b_f_std = 2
+
+# Draw from normal distribution to generate priors,
+# and add a bit of bias.
+a_f = rng.normal(loc=a_t + 0.5 * a_f_std, scale=a_f_std)
+b_f = rng.normal(loc=b_t + 0.5 * b_f_std, scale=b_f_std)
+
+def model(x, a, b):
+    return a * x + b
+
+obs_std = 0.5
+
+# Measurement matrix M is the linear model implemented as a matrix.
+M = np.ones(shape=(nparam, nparam))
+M[0, 0] = 0
+
+psi_f = np.array([a_f, b_f])
+C_f_psi = np.identity(nparam)
+C_f_psi[0, 0] = a_f_std ** 2
+C_f_psi[1, 1] = b_f_std ** 2
+
+C_eps = np.identity(nparam) * obs_std ** 2
+
+niter = 1
+
+for n in range(niter):
+    K = C_f_psi @ M.T @ np.linalg.inv(M @ C_f_psi @ M.T + C_eps)
+    d = np.array(
+        [
+            model(0, a_t, b_t) + rng.normal(loc=0, scale=obs_std),
+            model(1, a_t, b_t) + rng.normal(loc=0, scale=obs_std),
+        ]
+    )
+    psi_a = psi_f + K @ (d - M.dot(psi_f))
+    psi_a = psi_f + K @ (d - M.dot(psi_f))
+    C_a_psi = (np.identity(nparam) - K @ M) @ C_f_psi
+    psi_f = psi_a
+    C_f_psi = C_a_psi
+
+a_a = psi_a[0]
+b_a = psi_a[1]
+a_a_std = C_a_psi[0, 0]
+b_a_std = C_a_psi[1, 1]
+*/
+
+// Global variables used to test behaviour of series of estimates
+std::vector<double> a_diff_ml_posterior;
+std::vector<double> b_diff_ml_posterior;
+std::vector<double> covariance_determinant;
+SCENARIO("Running analysis update without row scaling on linear model",
+         "[analysis]") {
+
+    GIVEN("Different setups for analysis update") {
+        analysis_module_type *enkf_module = analysis_module_alloc("STD_ENKF");
+        // Default truncation of 0.98 of eigenvalue spectrum is such that this
```

Comment:
Perhaps add a sentence as to why?

Reply:
I've added a partial explanation, but this will have to be revisited.